core/types/rollup_cost.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2022 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package types
+
+import (
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+type RollupCostData struct {
+	zeroes, ones uint64
+}
+
+func NewRollupCostData(data []byte) (out RollupCostData) {
+	for _, b := range data {
+		if b == 0 {
+			out.zeroes++
+		} else {
+			out.ones++
+		}
+	}
+	return out
+}
+
+type StateGetter interface {
+	GetState(common.Address, common.Hash) common.Hash
+}
+
+// L1CostFunc is used in the state transition to determine the L1 data fee charged to the sender of
+// the transaction. It should *never* be called on Deposit transactions -- doing so will cause it
+// to cache the wrong DB state and may lead to chain forks.
+type L1CostFunc func(rcd RollupCostData, blockTime uint64) *big.Int
+
+// l1CostFunc is an internal version of L1CostFunc that also returns the gasUsed for use in
+// receipts.
+type l1CostFunc func(rcd RollupCostData) (fee, gasUsed *big.Int)
+
+var (
+	L1BasefeeSlot = common.BigToHash(big.NewInt(1))
+	OverheadSlot  = common.BigToHash(big.NewInt(5))
+	ScalarSlot    = common.BigToHash(big.NewInt(6))
+)
+
+var L1BlockAddr = common.HexToAddress("0x4200000000000000000000000000000000000015")
+
+// NewL1CostFunc returns a function used for calculating L1 fee cost.
+func NewL1CostFunc(config *params.ChainConfig, statedb StateGetter) L1CostFunc {
+	forBlock := ^uint64(0)
+	// The following variables are inintialized only *after* the first invocation of the returned
+	// function to ensure DB values are cached after deposit transactions have been processed.
+	var l1Basefee, overhead, scalar *big.Int
+	var isRegolith bool
+	var cachedFunc l1CostFunc
+	return func(rollupCostData RollupCostData, blockTime uint64) *big.Int {
+		if config.Optimism == nil {
+			return nil
+		}
```

Comment:
Previously it would return 0 when the rollup-gas is zero.

A zeroed `RollupCostData` can end up in an EVM message when `TransactionToMessage` is called on a deposit tx.
And even though it's not called for deposits in the real state-transition, `buyGas` still calls it, and RPC based EVM-calls need this.

Can we add back a case that returns 0, when the rollup-cost-data is equal to the empty `RollupCostData{}`?

Reply:
I see.  Added the empty check back in.